svg_loader: handling svgs without the viewBox attribs

mgrudzinska opened the pull request:
the first approach was made in #1248 but since many changes have been made since then,  a new pr is opened.

this solves #1239 and hopefully https://github.com/godotengine/godot/issues/72478

mgrudzinska commented on src/loaders/svg/tvgSvgLoader.cpp:
```diff
@@ -3294,6 +3317,8 @@ bool SvgLoader::read()
     TaskScheduler::request(this);
 
+    if (!isBBoxGiven) this->done();
+
```
@hermet @JSUYA 
this is an important change comparing to the #1248 
in case the isBBoxGiven == false, the vBox of the svg is known only after the whole file is handled.

hermet replied:
@mgrudzinska please remove the draft if you are ready to get a review.

mgrudzinska commented on src/loaders/svg/tvgSvgSceneBuilder.cpp:
```diff
@@ -788,3 +788,16 @@ unique_ptr<Scene> svgSceneBuild(SvgNode* node, float vx, float vy, float vw, flo
     return root;
 }
+
+void applyBBoxFlag(const Scene* scene, float& vx, float& vy, float& vw, float& vh, float& w, float& h, BBoxFlag bboxFlag)
```
The only reason I added this function is because I didn;t want to change the loader members (vx, vy, vw, vh) from the svgSceneBuild function. calling a dedicated function seems to be more readable, but still, it should be placed in some common space I think, but maybe you have some better idea in general how to deal with no viewBox info than bboxflag

JSUYA commented on src/loaders/svg/tvgSvgLoader.h:
```diff
@@ -51,6 +51,7 @@ class SvgLoader : public LoadModule, public Task
     unique_ptr<Paint> paint() override;
 
 private:
+    BBoxFlag bboxFlag;
```
How about initializing it to `BBoxFlag::None`?

JSUYA commented on src/loaders/svg/tvgSvgLoaderCommon.h:
```diff
@@ -146,6 +146,27 @@ enum class SvgParserLengthType
+    ViewboxH = 0x2,
+    Viewbox = ViewboxX | ViewboxY | ViewboxW | ViewboxH,
+};
+inline BBoxFlag operator|(BBoxFlag b1, BBoxFlag b2) {
+    return (BBoxFlag)((uint32_t)b1 | (uint32_t)b2);
+}
+inline BBoxFlag operator&(BBoxFlag b1, BBoxFlag b2) {
+    return (BBoxFlag)((uint32_t)b1 & (uint32_t)b2);
+}
+inline BBoxFlag& operator|=(BBoxFlag& b1, BBoxFlag b2) {
+    return (BBoxFlag&)((uint32_t&)b1 |= (uint32_t)b2);
+}
```
nice try! I don't know if it's appropriate to declare it here. Haha.. I expect that the readability will be much better if the existing masking codes are handled like this.

mgrudzinska replied:
removed for now -> todo in a separate pr

JSUYA commented on src/loaders/svg/tvgSvgSceneBuilder.cpp:
```diff
@@ -788,3 +788,16 @@ unique_ptr<Scene> svgSceneBuild(SvgNode* node, float vx, float vy, float vw, flo
+
+void applyBBoxFlag(const Scene* scene, float& vx, float& vy, float& vw, float& vh, float& w, float& h, BBoxFlag bboxFlag)
+{
+    if ((bboxFlag & BBoxFlag::Viewbox) != BBoxFlag::Viewbox) {
```
nit: BBoxFlag::Viewbox is full masking, so the code below is same.
`if (bboxFlag != BBoxFlag::Viewbox)`

mgrudzinska replied:
it was my bug, doesn;t exist after changes

JSUYA commented on src/loaders/svg/tvgSvgSceneBuilder.cpp:
```diff
@@ -788,3 +788,16 @@ unique_ptr<Scene> svgSceneBuild(SvgNode* node, float vx, float vy, float vw, flo
+        scene->bounds(nullptr, nullptr, &vw, &vh, false);
+        vx = 0.0f;
+        vy = 0.0f;
+        if ((bool)(bboxFlag & BBoxFlag::Width)) vw = w;
+        if ((bool)(bboxFlag & BBoxFlag::Height)) vh = h;
+    }
+    if (!(bool)(bboxFlag & BBoxFlag::Width)) w = vw;
+    if (!(bool)(bboxFlag & BBoxFlag::Height)) h = vh;
```
nit: It is recommended to use c++ style casts. (static_cast)

JSUYA reviewed (commented):
I'm not sure BBox is an appropriate name.
`BBoxFlag` enum is for masking the definition state(?) for width/height and viewbox.
And width/height and ViewboxW/ViewboxH are the same value, and there is no separate logic to distinguish them.
What do you think?
(I think there is no problem with the overall logic.)

mgrudzinska commented:
@JSUYA first of all I'm really very sorry for such a late answer. I have to read my code from the beginning and I guess you'll have to the the same... sorry

> I'm not sure BBox is an appropriate name. `BBoxFlag` enum is for masking the definition state(?) for width/height and viewbox. And width/height and ViewboxW/ViewboxH are the same value, and there is no separate logic to distinguish them. What do you think? (I think there is no problem with the overall logic.)

Thx for this note! since there was a bug - Width and ViewboxW shouldn't be the same. great that you've caught that :)
 
 I'll made changes asap (this week) and push them

mgrudzinska commented:
1.
> I'm not sure BBox is an appropriate name. `BBoxFlag` enum is for masking the definition state(?) for width/height and viewbox. 
I've changed the name BBox -> SvgViewFlag (since it has flags regarding both, the viewbox and the viewport).

2.
> And width/height and ViewboxW/ViewboxH are the same value, and there is no separate logic to distinguish them. What do you think? (I think there is no problem with the overall logic.)
It was a bug, they are not the same. Although I've changed the enum now - only if all of the viewBox params are correct the flag is set. 

3.
```
inline BBoxFlag operator|(BBoxFlag b1, BBoxFlag b2) {
    return (BBoxFlag)((uint32_t)b1 | (uint32_t)b2);
}
inline BBoxFlag operator&(BBoxFlag b1, BBoxFlag b2) {
    return (BBoxFlag)((uint32_t)b1 & (uint32_t)b2);
}
inline BBoxFlag& operator|=(BBoxFlag& b1, BBoxFlag b2) {
    return (BBoxFlag&)((uint32_t&)b1 |= (uint32_t)b2);
}
```
For now I used the same logic that we have for all other flags. I do agree that the readability has dropped, but the code is consistent now. This change I'll push in a separate pr and for all of the flags (or these that need this). 
Also I've used C-style casting, as it is for the other flags - separate pr, maybe after the above is applied.

mgrudzinska commented on src/loaders/svg/tvgSvgLoader.cpp:
```diff
@@ -831,20 +834,32 @@ static bool _attrParseSvgNode(void* data, const char* key, const char* value)
+            doc->viewFlag = (SvgViewFlag)((uint32_t)doc->viewFlag & ~(uint32_t)SvgViewFlag::Viewbox);
+            TVGLOG("SVG", "Negative values of the <viewBox> width and/or height - the attribute invalidated.");
+        }
+        if (!((uint32_t)doc->viewFlag & (uint32_t)SvgViewFlag::Viewbox)) {
```
cases like ```viewBox="0 0 200"``` - this should be ignored and wasn't before

mgrudzinska commented:
@hermet @JSUYA I hope that this time I covered in tests all of the cases...

[tests.tar.gz](https://github.com/thorvg/thorvg/files/10946876/tests.tar.gz)

results for master:
<img width="397" alt="Zrzut ekranu 2023-03-11 o 00 46 26" src="https://user-images.githubusercontent.com/67589014/224449139-f5828161-efb5-44ce-8453-1d9aa574640c.png">

results after applying this pr:
<img width="397" alt="Zrzut ekranu 2023-03-11 o 00 43 40" src="https://user-images.githubusercontent.com/67589014/224448896-6a8c6376-fd71-49e0-9dfb-cb0335d634b2.png">

hermet commented on src/loaders/svg/tvgSvgLoader.cpp:
```diff
@@ -3192,7 +3217,8 @@ void SvgLoader::run(unsigned tid)
     }
-    root = svgSceneBuild(loaderData.doc, vx, vy, vw, vh, w, h, align, meetOrSlice, svgPath);
+    root = svgSceneBuild(loaderData.doc, vx, vy, vw, vh, w, h, align, meetOrSlice, svgPath, viewFlag);
+    applySvgViewFlag(root.get(), vx, vy, vw, vh, w, h, viewFlag);
```
How about put this call in svgSceneBuild inside()? otherwise we don't need to expose the method.

mgrudzinska replied:
```if (!isBBoxGiven) this->done();``` - removed, also after changes it wasn;t necessary any more

applySvgViewFlag -> internal fun of the scene biulder, but now svgSceneBuild changes the vx, vy, vw, vh values

hermet reviewed (commented):
Thanks. please have a look at my comments.

capnm commented:
> the first approach was made in #1248 but since many changes have been made since then, a new pr is opened.
> 
> this solves #1239 and hopefully [godotengine/godot#72478](https://github.com/godotengine/godot/issues/72478)

There are still issues in images without the _width/height_ atribute, The change makes it recognized in applications like Godot, but it cannot be used because of the clipping. See https://github.com/godotengine/godot/issues/72478 

![texture_01](https://user-images.githubusercontent.com/4047289/224489879-2981bcd7-0b8e-4db6-87bb-d0783c5538a6.svg)
Gnome Image viewer :
![grafik](https://user-images.githubusercontent.com/4047289/224490019-fe552268-127d-49a6-8e3d-b264ba3227b3.png)

mgrudzinska commented on src/loaders/svg/tvgSvgLoader.cpp:
```diff
@@ -77,6 +77,9 @@
 #define PX_PER_CM 37.79528f //1 in = 2.54 cm -> PX_PER_IN/2.54
 
+#define DEFAULT_VIEWPORT_WIDTH  300.0f
+#define DEFAULT_VIEWPORT_HEIGHT 150.0f
```
it's not so simple - please check my last comment. I.ve changed the approach and removed these values.

hermet replied:
No one would not understand these values if these are out of standards. That's why I suggest you to leave any useful comments in the code for the maintanence - what brings you to use these specific values and if any fragiles when adjust the values.

mgrudzinska replied:
I removed them, since they seems to be a css std, not svg.

mgrudzinska commented on src/loaders/svg/tvgSvgLoader.cpp:
```diff
@@ -831,20 +834,32 @@ static bool _attrParseSvgNode(void* data, const char* key, const char* value)
+        }
+        if (!((uint32_t)doc->viewFlag & (uint32_t)SvgViewFlag::Viewbox)) {
+            loader->svgParse->global.x = loader->svgParse->global.y =
+            loader->svgParse->global.w = loader->svgParse->global.h = 0;
```
I've changed the def values: ```loader->svgParse->global.w = loader->svgParse->global.h = 1.0;``` in the ```_createSvgNode```. so if there is no reason to change them, they are already correct.

hermet replied:
ok

mgrudzinska commented on src/loaders/svg/tvgSvgLoader.cpp:
```diff
@@ -3174,6 +3194,11 @@ SvgLoader::~SvgLoader()
 
 void SvgLoader::run(unsigned tid)
 {
+    if (((uint32_t)viewFlag & (uint32_t)SvgViewFlag::Viewbox) && (fabsf(vw) < FLT_EPSILON || fabsf(vh) < FLT_EPSILON)) {
```
I moved the log to the header, but this check is still here, so I can return from the func. 
I don;t want to return false in a heder, since then the ```load``` api returns InvalidArguments.
is it what you meant?

hermet replied:
If you argue the condition is "Success" then it's ok otherwise InvalidArguments would be better or we can return some else.(you can define and suggest)

Regardless of that, there is no need to run a thread since we could know the condition prior to run() method.
Maybe you can check this at read() instead.

mgrudzinska replied:
I moved this check to read, now Result::Unknown is returned + log

mgrudzinska commented:
@hermet and @capnm as I wrote in #1248 : 
> 1. from https://www.w3.org/TR/SVG2/geometry.html#Sizing:
>    "The value auto for width and height on the svg element is treated as 100%."
> 2. from (for css) https://svgwg.org/specs/integration/#svg-css-sizing:
>    "When SVG element is the top level element, the [width](https://svgwg.org/svg2-draft/geometry.html#Sizing) and [height](https://svgwg.org/svg2-draft/geometry.html#Sizing) are set to '100%'
>    To resolve 'auto' value on ‘[svg](https://svgwg.org/svg2-draft/struct.html#SVGElement)’ element if the ‘[viewBox](https://svgwg.org/svg2-draft/coords.html#ViewBoxAttribute)’ attribute is not specified:
> 
> * Use author specified [width](https://svgwg.org/svg2-draft/geometry.html#Sizing) and [height](https://svgwg.org/svg2-draft/geometry.html#Sizing) ‘[svg](https://svgwg.org/svg2-draft/struct.html#SVGElement)’ element attributes.
> * If any of the sizing attributes are missing, resolve the missing ‘[svg](https://svgwg.org/svg2-draft/struct.html#SVGElement)’ element [width](https://svgwg.org/svg2-draft/geometry.html#Sizing) to '300px' and missing [height](https://svgwg.org/svg2-draft/geometry.html#Sizing) to '150px' (using [CSS 2.1 replaced elements size calculation](https://www.w3.org/TR/CSS21/visudet.html#inline-replaced-width)).
>   To resolve 'auto' value on ‘[svg](https://svgwg.org/svg2-draft/struct.html#SVGElement)’ element if the ‘[viewBox]"
> 
> 3. from https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/width#svg:
> Default value:	auto (treated as 100%)

so I think that the def values of the viewBox (300x150) aren't the svg standard, it's rather the css std. so I made some changes. it should be useful for godot now.

but:
- vx and vy are still 0 - I remember that there where cases in which this was a problem (shield and field svgs) - I need to read about that in the svg std one more time...
- for the UserSpaceOnUse coord sys some attribs have to be refreshed (like grads)

so after we decide the final version of this pr, I'll open (or not:) ) the issues regarding the above

hermet commented on src/loaders/svg/tvgSvgSceneBuilder.cpp:
```diff
@@ -755,21 +754,35 @@ static unique_ptr<Scene> _sceneBuildHelper(const SvgNode* node, const Box& vBox,
 }
 
 
+void _applySvgViewFlag(const Scene* scene, float& vx, float& vy, float& vw, float& vh, float& w, float& h, SvgViewFlag viewFlag)
```
static void

mgrudzinska replied:
done

mgrudzinska commented:
@hermet I still need to resolve this (I know what's happening, I'll solve this asap)
https://github.com/godotengine/godot/issues/72478#issuecomment-1465735893

hermet commented:
@mgrudzinska LGTM. please have @JSUYA confirm if possible. Also please squash 3 naugty commits instead rebase & merge. thanks!

mgrudzinska commented:
@hermet please taka a look on the last commit. I needed this to solve the issue described in (the size 1x1) https://github.com/godotengine/godot/issues/72478#issuecomment-1465735893

@JSUYA @capnm 
Please take a look and comment if necessary

hermet commented on src/lib/tvgLoadModule.h:
```diff
@@ -36,7 +36,7 @@ class LoadModule
     float vw = 0;
     float vh = 0;
-    float w = 0, h = 0;         //default image size
+    float w = 1.0f, h = 1.0f;         //default image size
```
You should not change the base policy for SVG. If necessary, please arrange the default values in SvgLoader.

mgrudzinska replied:
done

hermet commented on src/loaders/svg/tvgSvgLoader.cpp:
```diff
@@ -3292,6 +3314,9 @@ bool SvgLoader::read()
     if (!content || size == 0) return false;
 
+    if (((uint32_t)viewFlag & (uint32_t)SvgViewFlag::Viewbox) &&
+        (fabsf(vw) <= FLT_EPSILON || fabsf(vh) <= FLT_EPSILON)) return false;
```
IMHO, If this is a wierd case, we could print TVGLOG for hints.

mgrudzinska replied:
done

hermet reviewed (changes requested):
tvgLoadModule shouldn't be changed

hermet reviewed (approved):
LGTM thanks!

mgrudzinska commented:
@hermet the additional commit contains changes that we discussed, regarding width and height of the viewbox = 0. nothing changed in the previous one, that was already accepted

hermet reviewed (approved):
LGTM by me, also welcome @JSUYA's opinion.

JSUYA commented on src/lib/tvgPicture.cpp:
```diff
@@ -98,6 +98,7 @@ Result Picture::size(float w, float h) noexcept
 Result Picture::size(float* w, float* h) const noexcept
 {
     if (!pImpl->loader) return Result::InsufficientCondition;
+    pImpl->reload();
```
Should reload() called on size() and viewbox() getters?
Is there any overhead from this?

If the issue is the size when the user calls load() and immediately uses the size(), viewbox() getter
It seems that we can call resize() after reading in load() function.

mgrudzinska replied:
yes, if a user calls size/viewbox after load in a 'normal' case the information are already available, since header() was already called. but for svgs without width/height and viewbox info, this information is available after run() is finished, that's why I need to call reload here, to be sure that run is done.

JSUYA reviewed (approved):
I am so sorry tooooo late. 
I agree about the changes related to the loader's viewbox.
I left a comment regarding the reload() used for Picture class.
If there is no problem, please merge this. :)
